src/core/server/saved_objects/saved_objects_service.ts

```diff
@@ -22,40 +22,159 @@ import { first } from 'rxjs/operators';
+import { createRepository } from './service/lib/create_repository';
 import { Logger } from '..';
 
 /**
+ * Saved Objects is Kibana's data persisentence mechanism allowing plugins to
+ * use Elasticsearch for storing and querying state. The
+ * SavedObjectsServiceSetup API exposes methods for creating and registering
+ * Saved Object client wrappers.
+ *
+ * @remarks
+ * Note: The Saved Object setup API's should only be used for creating and
+ * registering client wrappers. Constructing a Saved Objects client or
+ * repository for use within your own plugin won't have any of the registered
+ * wrappers applied and is considered an anti-pattern. Use the Saved Objects
+ * client from the
+ * {@link SavedObjectsServiceStart | SavedObjectsServiceStart#scopedClient }
+ * method or the {@link RequestHandlerContext | route handler context} instead.
+ *
+ * When plugins access the Saved Objects client, a new client is created using
+ * the factory provided to `setClientFactory` and wrapped by all wrappers
+ * registered through `addClientWrapper`. To create a factory or wrapper,
+ * plugins will have to construct a Saved Objects client. First create a
+ * repository by calling `scopedRepository` or `internalRepository` and then
+ * use this repository as the argument to the {@link SavedObjectsClient}
+ * constructor.
+ *
+ * @example
+ * import {SavedObjectsClient, CoreSetup} from 'src/core/server';
+ *
+ * export class Plugin() {
+ *   setup: (core: CoreSetup) => {
+ *     core.savedObjects.setClientFactory(({request: KibanaRequest}) => {
+ *       return new SavedObjectsClient(core.savedObjects.scopedRepository(request));
+ *     })
+ *   }
+ * }
+ *
  * @public
  */
 export interface SavedObjectsServiceSetup {
-  clientProvider: ISavedObjectsClientProvider;
+  /**
+   * Set a default factory for creating Saved Objects clients. Only one client
+   * factory can be set, subsequent calls to this method will fail.
+   */
+  setClientFactory: (customClientFactory: SavedObjectsClientFactory<KibanaRequest>) => void;
+
+  /**
+   * Add a client wrapper with the given priority.
+   */
+  addClientWrapper: (
+    priority: number,
+    id: string,
+    factory: SavedObjectsClientWrapperFactory<KibanaRequest>
+  ) => void;
+
+  /**
+   * Creates a {@link ISavedObjectsRepository | Saved Objects repository} that
+   * uses the credentials from the passed in request to authenticate with
+   * Elasticsearch.
+   *
+   * @remarks
+   * The repository should only be used for creating and registering a client
+   * factory or client wrapper. Using the repository directly for interacting
+   * with Saved Objects is an anti-pattern. Use the Saved Objects client from
+   * the
+   * {@link SavedObjectsServiceStart | SavedObjectsServiceStart#scopedClient }
+   * method or the {@link RequestHandlerContext | route handler context}
+   * instead.
+   */
+  scopedRepository: (req: KibanaRequest, extraTypes?: string[]) => ISavedObjectsRepository;
```

Comment:
This should be a verb, `getScopedRepository`?

Reply:
renamed the repository methods into "create" since we're not just returning an existing repo but creating a new instance every time.